nengo_loihi/splitter.py

```diff
@@ -4,184 +4,280 @@
-        ) - self.passthrough.to_remove
+        # Forwards and backwards adjacency lists
+        pre_to_conn = defaultdict(list)
+        post_to_conn = defaultdict(list)
 
         # Initialize queue with the pre objects on host->chip connections.
         # We assume that all `conn.pre` objects are pre-computable, and then
         # raise an error later if one of them turns out to rely on chip output.
-        # Learning rules are not supported with precompute=True because
-        # this would require a hybrid simulation where some parts of the
-        # model interact with the host while other parts are precomputed
-        # ahead of time. The simulator assumes that precompute=True does not
-        # require any interaction between host and chip between time-steps.
-        # Also see issue #214.
-        for conn in conns:
+        for conn in self._conns:
             pre, post = base_obj(conn.pre), base_obj(conn.post)
             pre_to_conn[pre].append(conn)
             post_to_conn[post].append(conn)
             assert pre not in self.passthrough.to_remove
             assert post not in self.passthrough.to_remove
 
-            if isinstance(post, LearningRule) or conn.learning_rule is not None:
-                raise BuildError(
-                    "precompute=True not supported when using learning rules"
-                )
-
-            if self.on_chip(post) and not self.on_chip(pre):
-                mark_precomputable(pre)
+            if self.hostchip.on_chip(post) and not self.hostchip.on_chip(pre):
+                self.mark_precomputable(pre, queue)
 
-        # traverse all connected objects breadth-first
+        # Traverse all connected objects breadth-first
         while head < len(queue):
             node_or_ens = queue[head]
             head += 1
 
-            # handle forwards adjacencies
+            # Handle forwards adjacencies
             for conn in pre_to_conn[node_or_ens]:
                 assert base_obj(conn.pre) is node_or_ens
                 post = base_obj(conn.post)
-                if not self.on_chip(post):
-                    mark_precomputable(post)
+                if not self.hostchip.on_chip(post):
+                    self.mark_precomputable(post, queue)
 
-            # handle backwards adjacencies
+            # Handle backwards adjacencies
             for conn in post_to_conn[node_or_ens]:
                 assert base_obj(conn.post) is node_or_ens
                 pre = base_obj(conn.pre)
-                if self.on_chip(pre):
-                    raise BuildError(
-                        "Cannot precompute input, " "as it is dependent on output"
-                    )
-                mark_precomputable(pre)
+                if not self.hostchip.on_chip(pre):
+                    self.mark_precomputable(pre, queue)
+                else:
+                    # We found a closure! The model is not precomputable.
```

Comment:
Very minor: I wouldn't say that 'we found a closure'. In fact `pre` could just be some dangling ensemble on the chip that isn't receiving any input and isn't connected to anything else. We're doing the right thing in this case by detecting that it's not precomputable, since input to the chip (the first values in the queue) depend on output from the chip (the above `conn`). Think it's more accurate to say 'We found an input to the chip that relies on an output from the chip. The host model is not precomputable.'

Reply:
That makes sense, I changed it. Well, I say it makes sense, but it makes sense only after thinking real hard about it :dizzy_face: